sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala

```diff
@@ -211,15 +230,72 @@ class SliceArrowOutputProcessorImpl(
     batch
   }
 
-  override def loadBatch(): Boolean = {
-    if (rowCount > 0 && currentRowIdx < rowCount) {
-      true
+  protected override def getVectors(root: VectorSchemaRoot): Array[ColumnVector] = {
+    root.getFieldVectors.asScala.map { vector =>
+      new ArrowColumnVector(vector)
+    }.toArray[ColumnVector]
+  }
+
+  override def close(): Unit = {
+    if (prevRoot != null) {
+      prevVectors.foreach(_.close())
+      prevRoot.close()
+    }
+  }
+}
+
+class SliceBytesArrowOutputProcessorImpl(
+    reader: ArrowStreamReader,
+    pythonMetrics: Map[String, SQLMetric],
+    arrowMaxBytesPerOutputBatch: Int)
+  extends BaseSliceArrowOutputProcessor(reader, pythonMetrics) {
+
+  protected override def getRoot: VectorSchemaRoot = {
+    if (getBatchBytes(root) < arrowMaxBytesPerOutputBatch) {
+      currentRowIdx = currentRowIdx + root.getRowCount
+      root
     } else {
-      val loaded = super.loadBatch()
-      currentRowIdx = 0
-      loaded
+      var rootSlice = root
+      val stepRatio = 0.9
+      var sliceBatchCount = ((rowCount - currentRowIdx) * stepRatio).toInt
+
+      while (sliceBatchCount > 0) {
+        rootSlice = root.slice(currentRowIdx, sliceBatchCount)
+        val batchBytes = getBatchBytes(rootSlice)
+        if (batchBytes < arrowMaxBytesPerOutputBatch) {
+          currentRowIdx = currentRowIdx + rootSlice.getRowCount
+
+          return rootSlice
+        } else {
+          rootSlice.close()
+          sliceBatchCount = (sliceBatchCount * stepRatio).toInt
```

Comment:
So, simply, the number of rows are iteratively reduced by 10% until we meet the requirements? And, the worst case is to fallback to a single row at the line 277?

Reply:
Yes, we iteratively slice the batches until we reach a batch with required byte size. If we cannot have it even with one single record/row, we output a batch with one row (because as commented in the code, normally it should not happen but leave it for extreme case).